rust/src/main.rs

```diff
@@ -76,37 +69,147 @@ fn setup_logging() -> Result<(), fern::InitError> {
         }).level(level_filter)
         .chain(io::stderr());
 
-    let path_result = path_for_log_file(XI_LOG_FILE);
-    // If the logging_file_path returned successfully, add the logfile capability to fern
-    if let Ok(logging_file_path) = &path_result {
+    if let Some(logging_file_path) = &logging_path_result {
+        // Ensure the logging directory is created
+        let parent_path = logging_file_path.parent().ok_or(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!(
+                "Unable to get the parent of the following Path: {}",
+                logging_file_path.display(),
+            ),
+        ))?;
+        fs::create_dir_all(parent_path)?;
+        // Attach it to fern
         fern_dispatch = fern_dispatch.chain(fern::log_file(logging_file_path)?);
-    }
+    };
 
     // Start fern
     fern_dispatch.apply()?;
+    info!("Logging with fern is setup");
 
-    // If the logging_file_path returned an error, print it with fern/logs
-    if let Err(e) = &path_result {
-        let message = "There was an issue getting the path for the log file";
-        warn!("{}: {:?}, falling back to stderr.", message, e);
+    // Log details of the logging_file_path result using fern/log
+    // Either logging the path fern is outputting to or the error from obtaining the path
+    match &logging_path_result {
+        Some(logging_file_path) => info!(
+            "Logging to the following file: {}",
+            logging_file_path.display()
+        ),
+        None => warn!("There was no path supplied for the log file, falling back to stderr."),
     }
-
-    info!("Logging with fern is setup");
     Ok(())
 }
 
+fn generate_logging_path(logfile_config: LogfileConfig) -> Result<PathBuf, io::Error> {
+    // Use the file name set in logfile_config or fallback to the default
+    let logfile_file_name = match logfile_config.file {
+        Some(file_name) => file_name,
+        None => PathBuf::from(XI_LOG_FILE),
+    };
+    // Use the directory name set in logfile_config or fallback to the default
+    let logfile_directory_name = match logfile_config.directory {
+        Some(dir) => dir,
+        None => PathBuf::from(XI_LOG_DIR),
+    };
```

Comment:
Explicitly specifying an empty string would currently result in logging directly to `~/.local/` I think? Not sure that is the most intuitive thing to do... But I'm not sure what would be better either. Maybe default to the current working directory in that case?

Reply:
I ran it a couple ways and the current behavior is (for XDG style systems like Linux & Redox):
- `xi-core --log-dir` -> `/home/username/.local/share/xi-core/xi-core.log`
- `xi-core --log-dir --other-flag` -> `/home/username/.local/share/xi-core/xi-core.log`
- `xi-core --log-dir  --other-flag` -> `/home/username/.local/share/xi-core/xi-core.log` (2 spaces between the flags, doesn't seem to want to show up)
- `xi-core --log-dir ""` -> `/home/username/.local/share/xi-core.log`
- `xi-core --log-dir " "` -> `/home/username/.local/share/ /xi-core.log`
- `xi-core --log-dir something` -> `/home/username/.local/share/something/xi-core.log`
- `xi-core --log-dir "something"` -> `/home/username/.local/share/something/xi-core.log`

In my opinion those are as I'd expect them to be.

As for defaulting to the current directory, I was under the impression falling back to the current directory was part of the reason for this work as writing to within the `XiEditor.app/Contents/MacOS` folder was causing issues.
Also it's pretty common behavior for programs to go ahead and write logs off in.

Also assuming `/some/directory` was the working dir it's still easy to do with `$PWD` or an absolute path:

- `xi-core --log-dir .` -> `/home/username/.local/share/./xi-core.log`
- `xi-core --log-dir "."` -> `/home/username/.local/share/./xi-core.log`
- `xi-core --log-dir $PWD` -> `/some/directory/xi-core.log`
- `xi-core --log-dir /some/directory` -> `/some/directory/xi-core.log`
- `xi-core --log-dir "/some/directory"` -> `/some/directory/xi-core.log`
- `xi-core --log-dir ~/directory` -> `/home/username/directory/xi-core.log`
- `xi-core --log-dir "~/directory"` -> `/home/username/.local/share/~/directory/xi-core.log`

The only behavior that's a bit odd/annoying to me is:
Taking `"~/directory"` (in quotes) literally is quite normal cli behavior but `"/some/directory"` (in quotes) isn't taken in a similar way and results in `/some/directory/xi-core.log` instead of `/home/username/.local/share//some/directory/xi-core.log` (which would be simplified to `/home/username/.local/share/some/directory/xi-core.log`) as it starts with `/` and `PathBuf` sees that as a reason to replace the `PathBuf` with an absolute path.
I'd rather that be one way or the other.

Reply:
I tried `echo "~"` on `zsh` and `bash` on my machine and it didn't expand. `echo "$HOME"` and `echo ~` expand though.

Reply:
I think as it is now it all behaves in a predictable manner.
We can document it's behavior nice and clearly somewhere and the logging to stderr that says where the logfile is can be handy.


We could also add at some point a nice way for Frontends (& plugins) to see where the logfile is & show the user.
I can imagine bringing up a command pallet in a frontend and running a command called `Show logfile` or `Open logfile` or `Open logfile in file browser` or something.